### ISSUE TITLE
Critical bug - Read Request.Files - Stream it cause: Cannot access a closed file.

### ISSUE THREAD

**Issue by totpero** [Zaid-Ajaj/Nancy.Serilog#2]
Hi,

After i install this nuget into my project and activate serilog, i spent few hours to find what cause 
`Cannot access a closed file.`  when i try to access `Request.Files` and read archive Stream.

**Comment by Zaid-Ajaj** [comment]
Hi @totpero 
Can you please provide more information: 
 - Nancy version you using
 - Repro code

Are you sure the error is coming from this library? When I read the request body data, I also reset the stream position:
```csharp
public static string ReadBodyContent(this RequestStream inputStream)
{
    using (var streamReader = new StreamReader(inputStream))
    {
        var content = streamReader.ReadToEnd();
        // rewind stream to make it readable again from a Nancy module
        inputStream.Position = 0;
        return content;
    }
}
```

**Comment by totpero** [comment]
I will post sample app

**Comment by totpero** [comment]
This is demo app: [DemoNancySerilog.zip](https://www.mediafire.com/file/48j3qx6pwahs3k1/DemoNancySerilog.zip)
You can start in debug mode and put breakpoint  in HomeModule.cs on line 26 `file.Value.CopyTo(output);`
And using powershell you can run build.ps1 from project directory using cake build or you can try to post some files to http://localhost:60628/sync.

**Comment by totpero** [comment]
I published this demo project to this repo: https://github.com/totpero/DemoNancySerilog

**Comment by Zaid-Ajaj** [comment]
Thanks for the repro, I will take a look this evening

**Comment by Zaid-Ajaj** [comment]
Hi @totpero I have tried to reproduce your problem but for me it just works. I simplified your code a bit, I am reading the file contents in a StreamReader to get the data in memory and then process it however I want, please See the updated SelfhostedSample application:

```csharp
// public FilesModule() : base("/files")

Post["/sync"] = args =>
{
    var file = Request.Files.FirstOrDefault();
    if (file == null) return Response.AsText("No file provided");
    
    using (var reader = new StreamReader(file.Value))
    {
        var fileContent = reader.ReadToEnd();
        return Response.AsText($"File Contents: \n{fileContent}")
    }
};
``` 
Now in the `SelfhostedSample` I added a JSON file called `test-upload.json` with content:
```json
{
  "some-important": "data"
}
```
When running the application and sending this request:
```
cd SelfhostedSample
curl -F "test=@test-upload.json" http://localhost:4040/files/sync
```
I am then able to read the data properly 
![screenshot from 2018-02-22 10-39-10](https://user-images.githubusercontent.com/13316248/36530917-ab6b5bf2-17bc-11e8-9d9e-bbbeaddab13d.png)

So I am not entirely sure it has something to do with the library. Let me know what you think

**Comment by totpero** [comment]
Hi,
I update the repo, get last version and try the next steps:

1. Build the last project version;
2. Now you have two project (nancy web server with this nuget and console app that make `WebClient UploadFile`);
3. Run nancy web project (DemoNancySerilog) in debug mode;
4. In the same time go to `DemoNancySerilog.Console\bin\Debug` and run `DemoNancySerilog.Console.exe`
5. In the same folder with DemoNancySerilog.Console.exe app you will have "test.zip" file;
6. With your example code you will get `Cannot access a closed file` error:

![errr](https://user-images.githubusercontent.com/1710257/36564889-8049fe36-1827-11e8-8914-16edfdbaafc7.png)

My investigation result:

1. If you try to upload file using curl or powershell you don't get this error;
2. If you try to upload small file you don't get this error;
3. If you uninstall `Nancy.Serilog` nuget you don't get this error even if you have large file or upload using curl;

I hope this help you.

**Comment by Zaid-Ajaj** [comment]
Very interesting! thanks a lot for the update :smile: I am now finally able to reproduce the same error! 

The current updated version of Nancy.Serilog will lazy load log field values *only if they are not ignored*: this means if you ignore the `RequestBodyContent` then the library won't touch the request stream, you can use it for now to keep your application functional
```csharp
pipelines.EnableSerilog(new Options 
{
    IgnoredRequestLogFields = 
        Ignore.FromRequest()
              .Field(req => req.RequestBodyContent)
});
```
Configuring it this way means the function that reads the request body stream isn't invoked:
```cs
using (var streamReader = new StreamReader(inputStream))
{
        var content = streamReader.ReadToEnd();
        // rewind stream to make it readable again from a Nancy module
        inputStream.Position = 0;
        return content;
}
```
this means the problem is here somewhere, I need to investigate more and keep you updated

**Comment by Zaid-Ajaj** [comment]
Ok I did two things and published the changes in `0.8.0`:
1. Properly reset the stream reader when finished reading
2. Told the library *not to read the request body* if there are any files in there

1 is done like this:
```cs
public static string ReadBodyContent(this RequestStream inputStream)
{
    var content = "";
    using (var streamReader = new StreamReader(inputStream))
    {
        content = streamReader.ReadToEnd();
        // rewind stream to make it readable again from a Nancy module
        streamReader.DiscardBufferedData();
        streamReader.BaseStream.Seek(0, SeekOrigin.Begin);
    }
    return content;
}
```
2 is done like this:
```cs
if (!nancyRequest.Files.Any())
{
    // Only read request body content when there aren't any files
    request.RequestBodyContent = nancyRequest.Body.ReadBodyContent();
}
```
I think it makes more sense this way, because why would you log the content of the request files texually in a log field where the files usually contain binary data that is meant to be read as byte[].

If you want to log the files (or some of their metadata) you would have to do it manually via the contextual logger: `var logger = this.CreateLogger();`

Please update to 0.8.0 and let me know if it all goes smooth :smile:

**Comment by Zaid-Ajaj** [comment]
I believe this is fixed, so I will close the issue for now, feel free to open it again if the problem persists. :smile:
